cdc/sink/codec/canal_flat.go

```diff
@@ -0,0 +1,239 @@
+
+	return ret, nil
+}
+
+func (c *CanalFlatEventBatchEncoder) newFlatMessageForDDL(e *model.DDLEvent) (*canalFlatMessage, error) {
+	header := c.builder.buildHeader(e.CommitTs, e.TableInfo.Schema, e.TableInfo.Table, convertDdlEventType(e), 1)
+
+	ret := &canalFlatMessage{
+		ID:            0,
+		Schema:        header.SchemaName,
+		Table:         header.TableName,
+		IsDDL:         true,
+		EventType:     header.GetEventType().String(),
+		ExecutionTime: header.ExecuteTime,
+		BuildTime:     0,
+		Query:         e.Query,
+		tikvTs:        e.CommitTs,
+	}
+	return ret, nil
+}
+
+// EncodeCheckpointEvent is no-op
+func (c *CanalFlatEventBatchEncoder) EncodeCheckpointEvent(ts uint64) (*MQMessage, error) {
+	return nil, nil
+}
+
+// AppendRowChangedEvent implements the interface EventBatchEncoder
+func (c *CanalFlatEventBatchEncoder) AppendRowChangedEvent(e *model.RowChangedEvent) (EncoderResult, error) {
+	msg, err := c.newFlatMessageForDML(e)
+	if err != nil {
+		return EncoderNoOperation, errors.Trace(err)
+	}
+	c.unresolvedBuf = append(c.unresolvedBuf, msg)
+	return EncoderNoOperation, nil
+}
+
+// AppendResolvedEvent receives the latest resolvedTs
+func (c *CanalFlatEventBatchEncoder) AppendResolvedEvent(ts uint64) (EncoderResult, error) {
+	nextIdx := 0
+	for _, msg := range c.unresolvedBuf {
+		if msg.tikvTs <= ts {
+			c.resolvedBuf = append(c.resolvedBuf, msg)
+		} else {
+			break
+		}
+		nextIdx++
+	}
+	c.unresolvedBuf = c.unresolvedBuf[nextIdx:]
+	if len(c.resolvedBuf) > 0 {
+		return EncoderNeedAsyncWrite, nil
+	}
+	return EncoderNoOperation, nil
+}
+
+// EncodeDDLEvent encodes DDL events
+func (c *CanalFlatEventBatchEncoder) EncodeDDLEvent(e *model.DDLEvent) (*MQMessage, error) {
+	msg, err := c.newFlatMessageForDDL(e)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	value, err := json.Marshal(msg)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return NewMQMessage(nil, value, e.CommitTs), nil
+}
+
+// Build implements the EventBatchEncoder interface
+func (c *CanalFlatEventBatchEncoder) Build() []*MQMessage {
+	if len(c.resolvedBuf) == 0 {
+		return nil
+	}
+	ret := make([]*MQMessage, len(c.resolvedBuf))
+	for i := range c.resolvedBuf {
+		value, err := json.Marshal(c.resolvedBuf[i])
+		if err != nil {
+			log.Fatal("CanalFlatEventBatchEncoder", zap.Error(err))
+			return nil
+		}
+		ret[i] = NewMQMessage(nil, value, c.resolvedBuf[i].tikvTs)
+	}
+	c.resolvedBuf = c.resolvedBuf[0:0]
+	return ret
+}
+
+// MixedBuild is not used here
+func (c *CanalFlatEventBatchEncoder) MixedBuild(withVersion bool) []byte {
+	panic("MixedBuild not supported by CanalFlatEventBatchEncoder")
+}
+
+// Size implements the EventBatchEncoder interface
+func (c *CanalFlatEventBatchEncoder) Size() int {
+	return -1
+}
+
+// Reset resets the internal state of the encoder
+func (c *CanalFlatEventBatchEncoder) Reset() {
+	c.unresolvedBuf = make([]*canalFlatMessage, 0)
+	c.resolvedBuf = make([]*canalFlatMessage, 0)
+}
```

Comment:
`Reset` is only used in json encoder, is it better to leave a panic("not supported") only